unison-cli/src/Unison/Codebase/Editor/Slurp.hs

```diff
@@ -0,0 +1,534 @@
+  (Ord v, Show v) =>
+  Map (TaggedVar v) DefnStatus ->
+  Map (TaggedVar v) (Set (TaggedVar v)) ->
+  Map (TaggedVar v) (SummarizedStatus v)
+summarizeTransitiveStatus statuses deps =
+  flip imap (Align.align statuses deps) $ \tv -> \case
+    -- No dependencies
+    This selfStatus -> toSummary False selfStatus tv
+    That _set -> error $ "Encountered var without a status during slurping: " <> show tv
+    These selfStatus deps ->
+      let selfSummary = toSummary False selfStatus tv
+          summaryOfDeps = do
+            v <- Set.toList deps
+            depStatus <- maybeToList $ Map.lookup v statuses
+            pure $ toSummary True depStatus v
+       in foldl' pickPriorityStatus selfSummary summaryOfDeps
+  where
+    toSummary :: (Ord v, Show v) => Bool -> DefnStatus -> TaggedVar v -> SummarizedStatus v
+    toSummary isDep defNotes tv =
+      case defNotes of
+        DefOk Updated -> if isDep then NeedsUpdate tv else Ok Updated
+        DefErr err ->
+          if isDep
+            then ErrFrom tv err
+            else SelfErr err
+        DefOk New -> Ok New
+        DefOk Duplicated -> Ok Duplicated
+
+-- | Determine all variables which should be considered in analysis.
+-- I.e. any variable requested by the user and all of their dependencies,
+-- component peers, and component peers of dependencies.
+computeInvolvedVars ::
+  forall v.
+  Var v =>
+  UF.TypecheckedUnisonFile v Ann ->
+  Set v ->
+  Map (TaggedVar v) LD.LabeledDependency ->
+  Set (TaggedVar v)
+computeInvolvedVars uf defsToConsider varReferences
+  -- If nothing was specified, consider every var in the file.
+  | Set.null defsToConsider = Set.fromList $ Map.keys varReferences
+  | otherwise = varClosure uf requestedVarsWhichActuallyExist
+  where
+    -- The user specifies _untyped_ names, which may not even exist in the file.
+    -- We need to figure out which vars exist, and what type they are if they do.
+    requestedVarsWhichActuallyExist :: Set (TaggedVar v)
+    requestedVarsWhichActuallyExist = Set.fromList $ do
+      v <- Set.toList defsToConsider
+      -- We don't know whether each var is a type or term, so we try both.
+      tv <- [TypeVar v, TermVar v]
+      guard (Map.member tv varReferences)
+      pure tv
+
+-- | Compute transitive dependencies for all relevant variables.
+computeVarDeps ::
+  forall v.
+  Var v =>
+  UF.TypecheckedUnisonFile v Ann ->
+  Set (TaggedVar v) ->
+  Map (TaggedVar v) (Set (TaggedVar v))
+computeVarDeps uf allInvolvedVars =
+  let depMap :: (Map (TaggedVar v) (Set (TaggedVar v)))
+      depMap =
+        allInvolvedVars
+          & Set.toList
+          & fmap
+            ( \tv -> (tv, Set.delete tv $ varClosure uf (Set.singleton tv))
+            )
+          & Map.fromListWith (<>)
+   in depMap
+
+-- | Compute the closure of all vars which the provided vars depend on.
+-- A type depends on its constructors.
+varClosure :: (Var v) => UF.TypecheckedUnisonFile v a -> Set (TaggedVar v) -> Set (TaggedVar v)
```

Comment:
Can you describe what you mean by "shallow API"? I'm not familiar with that expression.

I agree that there's a bit of redundancy here though, some of that comes from maintaining backwards compatibility with SlurpResult and the associated machinery there. We can probably remove or isolate SlurpComponent if we want to, but I'd rather do that in a future PR and replace the Slurp Output stage of the operation (which we were already considering doing to show more precise slurping messages to users).

Agreed that we have a few types that are isomorphic to each other, the `TaggedVar`s are much easier to work with in this file because we often need to use vars as keys to maps, and since a TermVar has both TypeVar and TermVars in its dependencies it's tricky to work with when everything is a SlurpComponent.

My vote would be to leave it for now and revisit when we next work on the SlurpOutput, SlurpResult, and SlurpComponent sections 😄

Reply:
Yeah, my vote is to leave it now too.

By "shallow" I mean the newer API (`varClosure`, `partitionVars`, `mingleVars`) are small one-liners that wrap an existing API (`closeWithDependencies`). Generally deeper functions that encapsulate more functionality (and are thus more "worth" giving a name to) are preferred.